HDDS-4266: CreateFile : store parent dir entries into DirTable and file entry into separate FileTable

rakeshadr opened the pull request:
## What changes were proposed in this pull request?

https://issues.apache.org/jira/browse/HDDS-4266

## What is the link to the Apache JIRA
This task is to handle the createFileRequest. OM will store all the missing parent dirs to DirTable and leaf file node into a separate FileTable. This new FileTable and OpenFileTable is basically to segregate keys and files.

## How was this patch tested?

Added UTs - TestOMFileCreateRequestV1, TestOMKeyCommitRequestV1, TestOzoneFileOps.

linyiqun commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileCreateRequestV1.java:
```diff
@@ -0,0 +1,283 @@
+      // Add cache entries for the prefix directories.
+      // Skip adding for the file key itself, until Key Commit.
+      OMFileRequest.addDirectoryTableCacheEntries(omMetadataManager,
+              Optional.absent(), Optional.of(missingParentInfos),
```
I don't see we add OmDirectoryInfo in dir table cache during key commit. So could we put dir info into cache here?

linyiqun reviewed (commented):
@rakeshadr , thanks for working on this. Initial review comments for this. 
Haven't finished the review, will look deep on this.

linyiqun commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestV1.java:
```diff
@@ -0,0 +1,259 @@
+      if (!pathComponents.hasNext()) {
+        return lastKnownParentId;
+      }
+      String dbNodeName = lastKnownParentId + "/" + nodeName;
```
Can we use omMetadataManager.getOzonePathKey to construct db key name instead of?

rakeshadr replied:
Done!

linyiqun commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/file/OMFileCreateResponseV1.java:
```diff
@@ -0,0 +1,81 @@
+            getOmKeyInfo().getParentObjectID(), getOmKeyInfo().getFileName(),
+            getOpenKeySessionID());
+    omMetadataMgr.getOpenKeyTable().putWithBatch(batchOp, openKey,
+            getOmKeyInfo());
```
Seems we lack of below logic compared with original OMFileCreateResponse logic.
```java
    // update volume usedBytes.
    omMetadataManager.getVolumeTable().putWithBatch(batchOperation,
        omMetadataManager.getVolumeKey(omVolumeArgs.getVolume()),
        omVolumeArgs);
    // update bucket usedBytes.
    omMetadataManager.getBucketTable().putWithBatch(batchOperation,
        omMetadataManager.getBucketKey(omVolumeArgs.getVolume(),
            omBucketInfo.getBucketName()), omBucketInfo);
```

rakeshadr replied:
Done!

linyiqun commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/file/OMFileCreateResponseV1.java:
```diff
@@ -0,0 +1,81 @@
+/**
+ * Response for create file request layout version1.
+ */
+public class OMFileCreateResponseV1 extends OMFileCreateResponse {
```
We should add annotation for cleanup table.
```java
@CleanupTableInfo(cleanupTables = OPEN_FILE_TABLE)
```

rakeshadr replied:
Done!

linyiqun commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/file/TestOMFileCreateRequestV1.java:
```diff
@@ -0,0 +1,456 @@
+/**
+ * Tests OMFileCreateRequest V1 layout version.
+ */
+public class TestOMFileCreateRequestV1 extends TestOMKeyRequestV1 {
```
Suppose here logic is same with TestOMFileCreateRequest.
Can we fully reused the existed unit tests in TestOMFileCreateRequest?
Maybe that we can extend TestOMFileCreateRequest and override ozone configuration we set for V1 version .

rakeshadr replied:
Great idea. Done in the latest commit.

linyiqun reviewed (commented):
Below is the review comments for the unit tests.

linyiqun commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCommitRequestV1.java:
```diff
@@ -0,0 +1,379 @@
+/**
+ * Class tests OMKeyCommitRequestV1 class.
+ */
+public class TestOMKeyCommitRequestV1 extends TestOMKeyRequestV1 {
```
Similar comment above. Can we try to reuse the common logics here?

rakeshadr replied:
Done!

linyiqun commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileCreateRequestV1.java:
```diff
@@ -0,0 +1,283 @@
+      omClientResponse = new OMFileCreateResponseV1(omResponse.build(),
+          omFileInfo, missingParentInfos, clientID, omVolumeArgs, omBucketInfo);
+
+      result = Result.SUCCESS;
```
We should additionally increase dir created metric since we created missing parent dirs.

rakeshadr replied:
Done!

rakeshadr commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyInfo.java:
```diff
@@ -413,7 +461,8 @@ public KeyInfo getProtobuf(boolean ignorePipeline) {
         .setObjectID(objectID)
-        .setUpdateID(updateID);
+        .setUpdateID(updateID)
+        .setParentID(parentObjectID);
```
I am not persisting fileName, which is already the last name in the path component(keyName). I have added logic to prepare filename from Keyname. Hope this is fine?

linyiqun replied:
Yes, makes sense to me.

linyiqun replied:
Yes, makes sense to me.

rakeshadr commented:
Thanks a lot @linyiqun for the review comments and I have updated PR addressing the same. Please take another look at it when you get a chance.

bharatviswa504 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestV1.java:
```diff
@@ -0,0 +1,260 @@
+      String dbFileKey = omMetadataManager.getOzonePathKey(parentID, fileName);
+      dbOpenFileKey = omMetadataManager.getOpenFileName(parentID, fileName,
+              commitKeyRequest.getClientID());
+
```
checkDirectoryAlreadyExists check is missing which is added by HDDS-4155

rakeshadr replied:
Thanks for pointing out this. Good point. Done in latest commit.

linyiqun replied:
checkDirectoryAlreadyExists check is specific for enabling OzoneFileSystem, do we really need to apply this here?

bharatviswa504 replied:
Yes, when commitRequest is common for KeyCreate/FileCreate

bharatviswa504 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestV1.java:
```diff
@@ -0,0 +1,260 @@
+      }
+
+      bucketLockAcquired =
+              omMetadataManager.getLock().acquireLock(BUCKET_LOCK,
```
acquireLock -> acquireWriteLock. The old method is deprecated.

rakeshadr replied:
Good one. Done in latest commit

bharatviswa504 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestV1.java:
```diff
@@ -0,0 +1,260 @@
+    auditLog(auditLogger, buildAuditMessage(OMAction.COMMIT_KEY, auditMap,
+            exception, getOmRequest().getUserInfo()));
+
+    switch (result) {
```
Can we move this to a common method, and call the new method from both old version and this new version.

rakeshadr replied:
Sure.

bharatviswa504 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestV1.java:
```diff
@@ -0,0 +1,260 @@
+      omVolumeArgs.getUsedBytes().add(correctedSpace);
+      omBucketInfo.getUsedBytes().add(correctedSpace);
+
+      omClientResponse = new OMKeyCommitResponse(omResponse.build(),
```
OMKeyCommitResponse -> OMKeyCommitResponseV1

rakeshadr replied:
Good catch. Done in latest commit

bharatviswa504 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMKeyCommitResponseV1.java:
```diff
@@ -0,0 +1,61 @@
+                               @Nonnull OmBucketInfo omBucketInfo) {
+    super(omResponse, omKeyInfo, ozoneKeyName, openKeyName, omVolumeArgs,
+            omBucketInfo);
+    this.omKeyInfo = omKeyInfo;
```
These 3 are set, not used anywhere.

rakeshadr replied:
Sure. Done!

linyiqun commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java:
```diff
@@ -129,6 +132,10 @@
    * |----------------------------------------------------------------------|
+   * |  fileTable         | parentId/fileName/id -> FileInfo                |
+   * |----------------------------------------------------------------------|
+   * |  openFileTable     | parentId/fileName -> FileInfo                   |
```
openFileTable should have additional id after parentId/fileName and fileTable key should be parentId/fileName.

rakeshadr replied:
Thanks again. Done in latest commit

linyiqun commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/file/TestOMFileCreateRequest.java:
```diff
@@ -146,7 +141,7 @@ public void testValidateAndUpdateCache() throws Exception {
 
-    omKeyInfo = omMetadataManager.getOpenKeyTable().get(openKey);
+    omKeyInfo = verifyPathInOpenKeyTable(keyName, id, true);
     Assert.assertNotNull(omKeyInfo);
```
This check can be removed as we already check this in verifyPathInOpenKeyTable.

rakeshadr replied:
Done in latest commit

linyiqun reviewed (commented):
Thanks for addressing the comments, @rakeshadr . Few other comments from me.
Also please have a fix on the checkstyle warning.

rakeshadr commented on hadoop-ozone/interface-storage/src/main/java/org/apache/hadoop/ozone/om/OMMetadataManager.java:
```diff
@@ -399,4 +399,15 @@ String getMultipartKey(String volume, String bucket, String key, String
    * @return DB directory key as String.
    */
   String getOzonePathKey(long parentObjectId, String pathComponentName);
```
Done!

rakeshadr commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java:
```diff
@@ -129,6 +132,8 @@
    * |  directoryTable    | parentId/directoryName -> DirectoryInfo         |
    * |----------------------------------------------------------------------|
+   * |  fileTable         | parentId/fileName -> FileInfo                   |
+   * |----------------------------------------------------------------------|
```
Done!

rakeshadr commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/file/OMFileCreateResponseV1.java:
```diff
@@ -0,0 +1,81 @@
+    if (parentDirInfos != null) {
+      for (OmDirectoryInfo parentKeyInfo : parentDirInfos) {
+        String parentKey = omMetadataMgr.getOzonePathKey(
+                parentKeyInfo.getParentObjectID(), parentKeyInfo.getName());
```
Done!

rakeshadr commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java:
```diff
@@ -346,6 +369,7 @@ protected static DBStoreBuilder addOMTablesAndCodecs(DBStoreBuilder builder) {
         .addCodec(S3SecretValue.class, new S3SecretValueCodec())
         .addCodec(OmPrefixInfo.class, new OmPrefixInfoCodec())
         .addCodec(OmDirectoryInfo.class, new OmDirectoryInfoCodec())
+
```
Done!

rakeshadr commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileCreateRequestV1.java:
```diff
@@ -0,0 +1,283 @@
+
+      // add all missing parents to dir table
+      missingParentInfos = OMDirectoryCreateRequestV1.getAllParentDirInfo(
+              ozoneManager, keyArgs, pathInfoV1, trxnLogIndex);
```
Done!

rakeshadr commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMKeyCommitResponseV1.java:
```diff
@@ -0,0 +1,62 @@
+ */
+@CleanupTableInfo(cleanupTables = {OPEN_KEY_TABLE, KEY_TABLE,
+        OPEN_FILE_TABLE, FILE_TABLE})
+public class OMKeyCommitResponseV1 extends OMKeyCommitResponse {
```
Done!

rakeshadr commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileOps.java:
```diff
@@ -0,0 +1,206 @@
+  public Timeout timeout = new Timeout(300000);
+
+  private static final Logger LOG =
+          LoggerFactory.getLogger(TestOzoneDirectory.class);
```
Done!

rakeshadr commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileOps.java:
```diff
@@ -0,0 +1,206 @@
+            omMgr.getOpenKeyTable().iterator();
+    keysItr.seekToFirst();
+
+    // Assert.assertEquals("Unexpected directory table entries!", 2, count);
```
Done!

rakeshadr commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileOps.java:
```diff
@@ -0,0 +1,206 @@
+    }
+    Assert.assertEquals("Unexpected directory table entries!", 0, count);
+
+    // verify entries in directory table
```
Done!

rakeshadr commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyInfo.java:
```diff
@@ -126,6 +144,23 @@ public void setDataSize(long size) {
+  public long getParentObjectID() {
+    return parentObjectID;
+  }
+
```
Done!

rakeshadr commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileCreateRequestV1.java:
```diff
@@ -0,0 +1,289 @@
+                pathInfoV1.getLeafNodeName());
+        dbFileInfo = omMetadataManager.getKeyTable().get(dbFileKey);
+        if (dbFileInfo != null) {
+          ozoneManager.getKeyManager().refresh(dbFileInfo);
```
It makes sense to skip this refresh. FYI, I've followed existing code path. [Reference: ](https://github.com/apache/hadoop-ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileCreateRequest.java#L226)

bharatviswa504 replied:
Yes, checked that but we are writing new requests, if we can make it better that would be nice.

rakeshadr replied:
OK, got it. I will remove it in the next commit.

rakeshadr commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java:
```diff
@@ -227,6 +247,9 @@ protected OmMetadataManagerImpl() {
 
   @Override
   public Table<String, OmKeyInfo> getOpenKeyTable() {
+    if (enableFSPaths && OzoneManagerRatisUtils.isOmLayoutVersionV1()) {
```
The idea here is, will use the new tables only if enableFSPaths is true and with V1 version.

rakeshadr replied:
Good catch @bharatviswa504. Please feel free to add if anything else needed. Thanks again!

Based on our offline discussions, below is the expected behavior for diff requests:

**V1 feature version** : Following ops shouldn't depend on enableFSPaths flag
1) FileCreate  -----> Look into dirTable for parents. Then create entries in openFileTable and on close add it to fileTable.
2) DirCreate  -----> Create entries in dirTable
3) File/DirDelete -> Look into fileTable and dirTable for the keys.
4) File/DirRename-> Look into fileTable and dirTable for the keys.

**V1 feature version & enableFSPaths=true**
1) KeyCreate ---> Look into dirTable for parents. Create entries in openFileTable and on close add it to fileTable.
2) KeyDelete ---> Look into fileTable and dirTable for the keys.
3) KeyRename -> supported only in ozone shell. It should look into fileTable and dirTable for the keys.

**V1 feature version & enableFSPaths=false**
1) KeyCreate ---> Create entries in openKeyTable and on close add it to keyTable.
2) KeyDelete ---> Look into keyTable for the keys.
3) KeyRename -> supported only in ozone shell. It should look into keyTable for the keys.

In this PR, will handle only `FileCreate` request and not provided checks for enableFSPaths in KeyCommit. Will do this changes in latest commit.

Later, I will raise subsequent jiras for handling KeyCreate/KeyCommit and other ops.

linyiqun replied:
@rakeshadr , the V1 feature is a new key format and is not compatible with old format. enableFSPaths flag behaves as a switch here, why we remove this? Not so fully understand this.
Without enableFSPaths check, we will directly use new key format by default once V1 feature is supported.
Please correct me if I am wrong.

rakeshadr replied:
Thanks @linyiqun for the comment. It seems the above comment was not clear. I had done few corrections to the above comment. 

Yes, V1 represents new key format. 

Adding background about  `ozone.om.enable.filesystem.paths` -> this is the config to enable/disable enableFSPaths feature. Basically here the idea is to provide s3/fs inter-op. Please refer jira https://issues.apache.org/jira/browse/HDDS-4097 for more details. If the flag is enabled, then the user given key will be normalized and stored in FS semantics format by OM and it will be 100% FS semantics. If it is false, the key won't be normalized and it will be 100% S3 semantics.

For example, user created a key "/dir1/dir2/dir3/file1" from S3 API. Now,  if the flag is enabled the key will be normalized and create intermediate directories for the file1.

_**More Details:-**_
The cases I mentioned above - **V1 feature version & enableFSPaths=true** is 100% FS semantics and **V1 feature version & enableFSPaths=false** is 100% S3 semantics

Assume the key is /dir1/dir2/dir3/file-1. Again assume V1 feature version enabled and bucketId is 512.
Now, 
**enableFSPaths=true**, which is 100% FS semantics.
It stores as "512/dir1:1025", "1025/dir2:1026" and "1026/dir3:1027" into dirTable and "1027:file1" into openFiletable and on close move it to fileTable

**enableFSPaths=false**, which is 100% S3 semantics.
It stores as "512/dir1/dir2/dir3/file1:1025" into openFileTable and on close move it to fileTable. Here still maintains the parentID/Key format, but the key will be the fullpath and not a normalized path. Here the key can be anything like `/dir1////dir2////dir3///file1`.

Please let me know if any more details needed.

bharatviswa504 replied:
Just to add one more point here
this is FS API, so the usage of the flag is of no effect, as this is FS API (File create) not an Object Store API.

But KeyCommitRequest is common for both FS/Object Store we need special handling over there.

rakeshadr replied:
Sure, I will add special handling in KeyCommit code while implementing KeyCreate request. Hope that make sense to you.

bharatviswa504 replied:
Fine with me.

linyiqun replied:
Get it, Thanks @rakeshadr and @bharatviswa504 for the very clear explanation!

rakeshadr commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileCreateRequestV1.java:
```diff
@@ -0,0 +1,289 @@
+          .setOpenVersion(openVersion).build())
+          .setCmdType(Type.CreateFile);
+      omClientResponse = new OMFileCreateResponseV1(omResponse.build(),
+          omFileInfo, missingParentInfos, clientID, omVolumeArgs, omBucketInfo);
```
OMKeyInfo is exposed to Ozone client, so I have kept 'keyName' as it is. Anyway fileName can be extracted from the  keyName so that I thought of reducing the load on DB.

Hope this make sense to you.

bharatviswa504 replied:
But now OmKeyInfo has entire keyPath which is not really needed to be persisted to DB in this new V1 Request.

rakeshadr replied:
I completely agree with you that, V1 need only the fileName in DB.

On a second thought, do we need to maintain two different `OMKeyInfo#keyName` semantics in DB. For 100% S3 , 'keyName' stores full path. For 100% FS, 'keyName' stores only the fileName in the value part. 
```
message keyInfo {
       ....
       required string keyName = 3;
```

bharatviswa504 replied:
Yes. Because if we store the entire keyPath in DB when rename, then if we change the only parentObjectID and then leave keyName then they will diverge.

And also this is FS API not called from S3.

rakeshadr replied:
Fully agreed with you, thanks for pointing out the rename case. 

How about store `fileName` in the `keyName` field in DB. Then, while populating OmKeyInfo logically will prepare fullPath in `keyName` and will store `fileName` separately.

bharatviswa504 replied:
Yes, you read my mind :)
It works perfectly.

rakeshadr replied:
Done the changes in new commit

bharatviswa504 replied:
Added to the cache correctly(With KeyName as leaf node name), but in response class, we passed omKeyInfo this should be also the same thing which we have added to the cache right?

rakeshadr replied:
Inside OMFileCreateResponseV1#addToDBBatch() method has the actual call to the DB, it is doing the changes in utility OMFileRequest function. I have done this way to ensure that, only when updating the TableCache and persisting to the Table it makes the DB representation as `keyName=leafNodeName`. All the other places(any functional checks inside OM, logging inside OM, client response etc..) will have the `keyName=fullPath` and `fileName=leafNodeName`.

   ```
 OMFileRequest.addToOpenFileTable(omMetadataMgr, batchOp, getOmKeyInfo(),
            getOpenKeySessionID());
```

Also, for better code maintenance, I will be adding these conversion in `OMFileRequest` utility functions. Hope this is fine for you?

bharatviswa504 replied:
With the current approach in PR looks like we do unnecessarily one more time copy object. What we add to the cache, if we pass the same instance to Response class then same can be persisted. With this way cache and Response classes uses the same instance, and anyway in logging in createRequest we use Keyname, I see no issue over there.

rakeshadr replied:
Sure, updated patch addressing this point. Thanks!

rakeshadr commented:
Thanks a lot @bharatviswa504 and @linyiqun for the review efforts and useful comments. I've made another commit by addressing your comments.

linyiqun commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:
```diff
@@ -253,4 +253,33 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
     return omClientResponse;
   }
+
+  protected void processResult(CommitKeyRequest commitKeyRequest,
```
Seems we forget to reuse processResult in OMKeyCommitRequest.

rakeshadr replied:
oops, thanks a lot @linyiqun  for the detailed code reviews. I have fixed this in my latest commit.

linyiqun reviewed (commented):
@rakeshadr , thanks for addressing all the comments!

rakeshadr commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileCreateRequestV1.java:
```diff
@@ -0,0 +1,289 @@
+          .collect(Collectors.toList());
+      omFileInfo.appendNewBlocks(newLocationList, false);
+
+      omVolumeArgs = getVolumeInfo(omMetadataManager, volumeName);
```
I've raised HDDS-4321 jira to track the changes and update to the branch.

bharatviswa504 replied:
Thanks.

rakeshadr commented:
@linyiqun I've resolved the comments which are fixed. Please feel free to re-open, if anything extra to be handled.

linyiqun commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestV1.java:
```diff
@@ -0,0 +1,275 @@
+              getDirectoryTable().get(dbNodeName);
+      if (omDirectoryInfo != null) {
+        checkDirectoryAlreadyExists(keyName, ozoneManager,
+                reachedLastPathComponent);
```
```xml
  protected boolean checkDirectoryAlreadyExists(String volumeName,
      String bucketName, String keyName, OMMetadataManager omMetadataManager)
      throws IOException {
    if (omMetadataManager.getKeyTable().isExist(
        omMetadataManager.getOzoneDirKey(volumeName, bucketName,
            keyName))) {
      return true;
    }
    return false;
  }
```
Original logic is to check if there is the same name dir key already stored in key table.
For here,  the logic is that we should use new file key to check if that existed in file table, not check the dir table.
checkDirectoryAlreadyExists will always throw OMException error once OzoneFileSystem enabled under above logic, because parents dir is already created during File create request.

rakeshadr replied:
Thanks for the comment. I think the logic is fine, but will discuss for better clarity and happy to fix if we finds any gap.

Yes, the existing logic in `OMKeyRequest#checkDirectoryAlreadyExists` checks the directory existence for the given committing fileName.

New logic also does the directory existence check against `DirTable` for the given committing fileName. `FileTable` stores only the file information <parentID/fileName> and will not know about the DirInfo.

For example,  committing fileName is `/dir1/dir2/dir3/file1`.

**Existing Logic:** 
step-1) It will add trailing slash - `/dir1/dir2/dir3/file1/` as the directory key entries represented by trailing slash.
step-2) Then check `/dir1/dir2/dir3/file1/` in the KeyTable. 

**New Logic:** 
step-1) Traverse till the leaf node. I have used a flag `boolean reachedLastPathComponent = !pathComponents.hasNext();` to mark that the traversal reached till end.
step-2) Once it reached the last path element, if it finds an OmDirectoryInfo from `DirTable` corresponding to the 'file1' then it throws OMException.
```

    omDirectoryInfo = omMetadataManager.
              getDirectoryTable().get(dbNodeName);
      if (omDirectoryInfo != null) {
        checkDirectoryAlreadyExists(keyName, ozoneManager,
                reachedLastPathComponent);
        lastKnownParentId = omDirectoryInfo.getObjectID();
      } 
```

linyiqun replied:
Thanks for the detailed explanation, I missed that fileName can also be passed during directory key entries check.

linyiqun reviewed (commented):
Please see my latest comments, @rakeshadr .
Other places seem good to me now.

rakeshadr commented:
Thank you @linyiqun and @bharatviswa504  for the continuous help in reviews. Please let me know if any more comments.

bharatviswa504 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java:
```diff
@@ -453,4 +456,145 @@ public static void addDirectoryTableCacheEntries(
+      // New key format for the openFileTable.
+      // For example, the user given key path is '/a/b/c/d/e/file1', then in DB
+      // keyName field stores only the leaf node name, which is 'file1'.
+      OmKeyInfo dbOmFileInfo = omFileInfo.copyObject();
```
Looks for every key commit now we do 2 times copy once to get from the table and here.
I understand the reason.

Nothing needs to be done here, just mentioning the difference between the original request and V1.

rakeshadr replied:
OK:-)

rakeshadr commented:
> Thank you @linyiqun and @bharatviswa504 for the continuous help in reviews. Please let me know if any more comments.

@linyiqun, @bharatviswa504  Can you please approve the PR if there are no more comments. Thanks again!

linyiqun reviewed (approved):
+1 from me. Thanks @rakeshadr for working on this.

rakeshadr commented:
> +1 from me. Thanks @rakeshadr for working on this.

Thank you for the quick response and your review comments helped a lot to cover many cases.

bharatviswa504 commented:
Thank You for the updated patch.
I have one comment, (resolved fixed comments), rest LGTM.

rakeshadr commented:
> Thank You for the updated patch.
> I have one comment, (resolved fixed comments), rest LGTM.

Thanks @bharatviswa504 for the detailed reviews. I've updated the patch addressing the comment.

bharatviswa504 reviewed (approved):
+1 LGTM
Thank You @rakeshadr for the contribution and for continuously updating the patch patiently and @linyiqun for the review.

rakeshadr commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java:
```diff
@@ -466,27 +465,22 @@ public static void addDirectoryTableCacheEntries(
    * @param trxnLogIndex      transaction log index
    * @return dbOmFileInfo, which keeps leaf node name in keyName field
```
Noted. Will take care in next PR.
